### ISSUE TITLE
[no-invalid-this] Probably false positive for callback

### ISSUE THREAD

**Issue by ilyub** [typescript-eslint/typescript-eslint#3620]
**Repro**

```TS
class MyClass {
  public method(): void {}
}

function f(callback: (this: MyClass) => void): void {
  callback.call(new MyClass());
}

f(function () {
  this.method();
});
```

**Expected Result**

This code should pass, because anonymous function is immediately consumed as an argument of "f", and "f" defines this for its argument. So, this is expected here, rather than invalid.

**Actual Result**

The rule shows error.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.28.1` |
| `@typescript-eslint/parser`        | `4.28.1` |
| `TypeScript`                       | `4.3.5` |
| `ESLint`                           | `7.30.0` |
| `node`                             | `14.17.0` |

**Comment by bradzacher** [comment]
this rule is intentionally simple and does not do any type analysis.
this is working as intended.
